README
moviedb

This simple app that implement a storage system for movies.
The interface is command line.
Movie information stored persistently in database(like SQLite, MySQL, etc).
For more information see perldoc App/moviedb.pm


DEPENDENCIES

perl v5.10+
Config::Tiny
DBI
DBD::SQLite / maybe another in future
IO::Prompt
Try::Tiny


INSTALLATION

To install this app, run the following commands:

	perl Makefile.PL
	make
	make test
	make install (currently not supported)


SETUP

Create database
  for SQLite like:
    $ sqlite3 data/moviedb.sqlite < conf/moviedb.schema.sql
  for MySQL like(not tested yet):
    $ mysqladmin create moviedb
    $ mysql -e "GRANT ALL PRIVILEGES ON moviedb.* TO moviedb@localhost IDENTIFIED BY 'password'"
    $ mysql -umoviedb -ppassword moviedb < conf/moviedb.schema.sql
    $ mysql -umoviedb -ppassword moviedb -e "ALTER TABLE movie MODIFY movie_id SERIAL; ALTER TABLE star MODIFY star_id SERIAL;"

Copy conf/moviedb.conf.example to conf/moviedb.conf
  and edit it


USE

Run bin/moviedb and analyze output


COPYRIGHT AND LICENCE

Copyright (C) 2011, Sergiy Borodych

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.


Changes
Revision history for App-movie_db

0.0.1  Sat Aug 13 23:14:50 2011
       Initial release.



Makefile.PL

use 5.010;
use strict;
use warnings;

use ExtUtils::MakeMaker;

WriteMakefile(
    NAME            => 'App::moviedb',
    AUTHOR          => 'Sergiy Borodych <bor@cpan.org>',
    VERSION_FROM    => 'lib/App/moviedb.pm',
    ABSTRACT_FROM   => 'lib/App/moviedb.pm',
    EXE_FILES       => [ 'bin/moviedb' ],
    PREREQ_PM       => {
        'Config::Tiny'  => 0,
        'DBI'           => 0,
        'DBD::SQLite'   => 0,
        'IO::Prompt'    => 0,
        'Try::Tiny'     => 0,
    },
    dist            => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean           => { FILES => 'App-moviedb-*' },
);


# using perl v5.10 is a bad idea though (it not supported and now we have v5.14 already)
warn <<'EOF' if $] <= 5.012;

*******************************************************************************
You are using Perl 5.10, a version that has reached the end of its life
and which is no longer getting updated by the community.

Therefore we highly recommend that you upgrade as soon as possible!
*******************************************************************************

EOF


#sub MY::postamble {
#    my $self = shift;
#    return <<"TEXT"
#TEXT
#}



lib/App/moviedb.pm
package App::moviedb;

use 5.010;
use lib qw( lib );
use warnings;
use strict;

our $VERSION = 0.01;

use Carp;
use Config::Tiny;
use DBI;
use DBD::SQLite;
use FindBin;
use Getopt::Long;
use Try::Tiny;

use App::moviedb::Command;

sub new {
    my $class = shift;
    my $self = bless {}, $class;
	# be sure to load config first
    try { $self->conf() } catch { die "ERROR: $_\n" };
    $self->{_dbh} = try { App::moviedb::DB->new( $self->conf('db') ); }
    				catch { die "ERROR: $_\n" };
    return $self;
}

sub run {
    my $self = shift;

    # process args
    GetOptions(
        'action|a=s' => \$self->{opt}{act},
        'config|c=s' => \$self->{opt}{conf_file},
        'file|f=s'   => \$self->{opt}{file_in},
        'help|h'     => \$self->{opt}{help},
    ) or die $self->usage();
    die $self->usage() if $self->{opt}{help};
    die die "ERROR: Require some args\n" . $self->usage()
      unless $self->{opt}{act};

    # call command
    my $command = App::moviedb::Command->new();
    my $method = 'act_' . delete $self->{opt}{act};
    if ( $command->can($method) ) {
        my $result =
          try { $command->$method( $self->{opt} ); }
          catch { die "ERROR: $_\n" . $self->usage(); };
        say $result if $result;
    }
    else {
        die "ERROR: Bad action name\n" . $self->usage();
    }
    return 1;
}

sub usage {
    return <<"USAGE";
Usage:
 $0 [-c path/to/config ] -a action_name [-f filename]
    -a      - do this action_name (some actions need additional params)
    -c      - using alternative config, default: conf/moviedb.conf
    -f      - use this file (for import ex.)
    -h      - show this screen
  where action_name:
    add             - add movie
    del             - delete movie
    display         - display movie (show properties)
    find            - find movies by title
    find_by_star    - find movies by star
    import          - import movies from text file
    list            - list movies by title
    list_by_year    - list movies by year
USAGE
}

# return: all config hashref or some part of config
sub conf {
    my ( $self, $what ) = @_;
    # load config if need
    unless ( $self->{_conf} ) {
    	$self->{_conf_dir} ||= "$FindBin::Bin/../conf";
    	my $conf_file = $self->{opt}{conf_file}
    	  || $self->{_conf_dir} . '/moviedb.conf';
    	if ( -e $conf_file ) {
    	    $self->{_conf} = Config::Tiny->read($conf_file)
    	      or die "Cant load config file $conf_file: " . Config::Tiny->errstr;
    	}
    	else {
    	    die "Cant find config file $conf_file\n";
    	}
	}
    return $what ? $self->{_conf}{$what} : $self->{_conf};
}

1;

__END__

=head1 NAME

App::moviedb - simple movie database

=head1 SYNOPSIS

    use App::moviedb;
    my $app = App::moviedb->new();
    $app->run();

=head1 DESCRIPTION

This simple app that implement a storage system for movies.
The interface is command line.
Movie information stored persistently in database(like SQLite, MySQL, etc).

=head1 CONFIGURATION AND ENVIRONMENT

App::moviedb requires configuration file for working.
Usually it placed in conf/moviedb.conf.

=head1 DEPENDENCIES

perl v5.10+,
Config::Tiny,
DBI,
DBD::SQLite / maybe another in future,
IO::Prompt,
Try::Tiny

=head1 AUTHOR

Sergiy Borodych  C<< <bor@cpan.org> >>

=head1 LICENCE AND COPYRIGHT

Copyright (c) 2011, Sergiy Borodych C<< <bor@cpan.org> >>. All rights reserved.

This module is free software; you can redistribute it and/or
modify it under the same terms as Perl itself. See L<perlartistic>.


lib/App/moviedb/Star.pm
package App::moviedb::Star;

use 5.010;
use lib qw( lib );
use warnings;
use strict;

our $VERSION = 0.01;

use App::moviedb::DB;

sub new {
    my ( $class, $params ) = @_;
    $params->{dbh} ||= App::moviedb::DB->new()->dbh();
    my $info =
      $params->{dbh}->selectrow_hashref(
        'SELECT star_id, name FROM star WHERE name = ?',
        undef, $params->{name}
      );
    return unless $info;
    my $self = bless $info, $class;
    $self->{dbh} = delete $params->{dbh};
    return $self;
}

1;

__END__

=head1 NAME

App::moviedb::Star

=head1 SYNOPSIS

    use App::moviedb::Star;
    my $movie = App::moviedb::Star->new({ name => $star_name, dbh => $dbh });
    say $movie->as_string();

=head1 DESCRIPTION

Simple implementation for star object.

=head1 SEE ALSO

App::moviedb

=head1 AUTHOR

Sergiy Borodych  C<< <bor@cpan.org> >>

=head1 LICENCE AND COPYRIGHT

Copyright (c) 2011, Sergiy Borodych C<< <bor@cpan.org> >>. All rights reserved.

This module is free software; you can redistribute it and/or
modify it under the same terms as Perl itself. See L<perlartistic>.


lib/App/moviedb/Movie.pm
package App::moviedb::Movie;

use 5.010;
use lib qw( lib );
use warnings;
use strict;

our $VERSION = 0.01;
our @EXPORT_OK = qw( %fields %field_titles );

use App::moviedb::DB;
use App::moviedb::Star;

our %fields = (
    # re - is like 'require' flag for IO::Prompt
    title  => { title => 'Title',        re => qr/^[\w\s\-:,]{1,255}$/ },
    year   => { title => 'Release Year', re => qr/^\d{4}$/ },
    format => { title => 'Format',       re => [qw( VHS DVD Blu-Ray )] },
    stars  => { title => 'Stars',        re => qr/^(?:[\w\s]+,?)+$/ },
);
our %field_titles = ( map { $fields{$_} => $_ } keys %fields );

sub new {
    my ( $class, $params ) = @_;
    die "Require movie id\n" unless $params->{movie_id};
    $params->{dbh} ||= App::moviedb::DB->new()->dbh();
    my $movie_info =
      $params->{dbh}->selectrow_hashref(
        'SELECT movie_id, title, year, format FROM movie WHERE movie_id = ?',
        undef, $params->{movie_id}
      );
    return unless $movie_info;
    my $self = bless $movie_info, $class;
    $self->{dbh} = delete $params->{dbh};
    return $self;
}

# TODO : move all DB related stuff to ::DB module
# add movie to DB, class method
# return: self (movie object)
sub add {
    my ( $class, $params ) = @_;
    my $self = bless $params, $class;
    $self->{dbh} ||= App::moviedb::DB->new()->dbh();
    # TODO : check if movie (title&year) already in DB
    # add movie info to DB
    $self->{dbh}->do(
        'INSERT INTO movie (title, year, format) VALUES (?,?,?)',
        undef, @{$self}{qw(title year format)}
    );
    $self->{movie_id} = $self->{dbh}->last_insert_id( undef, undef, 'movie', 'movie_id' );
    # add stars & movie_stars to DB
    my $sth = $self->{dbh}->prepare('INSERT INTO star (name) VALUES (?)');
    my $sth_ms = $self->{dbh}->prepare('INSERT INTO movie_star (movie_id, star_id) VALUES (?,?)');
    foreach my $name ( split(/,\s*/,$params->{stars}) ) {
        # check if already star exists
        my $star = App::moviedb::Star->new({ name => $name });
        unless ( $star ) {
            $sth->execute($name);
            $star->{id} = $self->{dbh}->last_insert_id( undef, undef, 'star', 'star_id' );
        }
        $sth_ms->execute($self->{movie_id},$star->{star_id});
    }
    $sth->finish();
    $sth_ms->finish();
    return $self;
}


# return: string with movie details
sub as_string {
    my $self = shift;
    my $stars = join( ', ', @{ $self->stars() } );
    return <<"MOVIE_INFO";

ID:     $self->{movie_id}
Title:  $self->{title}
Year:   $self->{year}
Format: $self->{format}
Stars:  $stars
MOVIE_INFO
}

# return: arrayref with star's names
sub stars {
    my $self = shift;
    unless ($self->{_stars}) {
        $self->{_stars} = $self->{dbh}->selectcol_arrayref(
            q/  SELECT star.name
                FROM star
                JOIN movie_star USING (star_id)
                WHERE movie_id = ?
            /,
            undef, $self->{movie_id}
        );
    }
    return $self->{_stars};
}

1;

__END__

=head1 NAME

App::moviedb::Movie

=head1 SYNOPSIS

    use App::moviedb::Movie;
    my $movie = App::moviedb::Movie->new({ id => $movie_id });
    say $movie->as_string();

=head1 DESCRIPTION

Simple implementation for movie object.

=head1 SEE ALSO

App::moviedb

=head1 AUTHOR

Sergiy Borodych  C<< <bor@cpan.org> >>

=head1 LICENCE AND COPYRIGHT

Copyright (c) 2011, Sergiy Borodych C<< <bor@cpan.org> >>. All rights reserved.

This module is free software; you can redistribute it and/or
modify it under the same terms as Perl itself. See L<perlartistic>.


lib/App/moviedb/Movies.pm
package App::moviedb::Movies;

use lib qw( lib );
use warnings;
use strict;

our $VERSION = 0.01;

# input: expect arrayref as $params
sub new {
    my ( $class, $params ) = @_;
    my $self = bless $params, $class;
    return $self;
}

# return: string with movie's info
sub as_string {
    my $self = shift;
    my $result;
    foreach my $movie ( @{ $self } ) {
        $result .= "ID: $movie->{movie_id}\tTitle: $movie->{title}\n";
    }
    $result = "\n" . $result if $result;
    return $result;
}

1;

__END__

=head1 NAME

App::moviedb::Movies

=head1 SYNOPSIS

    use App::moviedb::Movies;
    my $movies = App::moviedb::Movies->new(\@movies);
    say $movies->as_string();

=head1 DESCRIPTION

Simple implementation for display result for collection of movies.

=head1 SEE ALSO

App::moviedb

=head1 AUTHOR

Sergiy Borodych  C<< <bor@cpan.org> >>

=head1 LICENCE AND COPYRIGHT

Copyright (c) 2011, Sergiy Borodych C<< <bor@cpan.org> >>. All rights reserved.

This module is free software; you can redistribute it and/or
modify it under the same terms as Perl itself. See L<perlartistic>.


lib/App/moviedb/Command.pm
package App::moviedb::Command;

use 5.010;
use lib qw( lib );
use warnings;
use strict;

our $VERSION = 0.01;

use IO::Prompt;

use App::moviedb::DB;
use App::moviedb::Movie;
use App::moviedb::Movies;

sub new {
    my ( $class, $params ) = @_;
    $params //= {};
    my $self = bless $params, $class;
    $self->{db} ||= App::moviedb::DB->new();
    return $self;
}

# actions methods
#   with 'act_' prefix

# add movie to DB
sub act_add {
    my $self = shift;
    my %info;
    foreach my $field (qw( title year format stars )) {
        prompt(
            "$App::moviedb::Movie::fields{$field}{title}: ",
            -require => {
                "(bad value)\n$App::moviedb::Movie::fields{$field}{title}: " =>
                  $App::moviedb::Movie::fields{$field}{re}
            }
        );
        $info{$field} = $_;
    }
    my $movie = App::moviedb::Movie->add( \%info );
    return $movie ? 'Added successfully' : 'Cant add: something wrong';
}

# delete movie from DB
sub act_del {
    my $self = shift;
    prompt( 'ID: ', '-integer' );
    my $result = $self->{db}->del_movie($_);
    return $result ? 'Deleted successfully' : 'Cant delete: something wrong';
}

# display movie (show properties)
sub act_display {
    my $self = shift;
    prompt( 'ID: ', '-integer' );
    my $movie = App::moviedb::Movie->new({ movie_id => $_ });
    return $movie ? $movie->as_string() : 'None found';
}

# find movies by title
sub act_find {
    my $self = shift;
    while ( ! prompt( 'Title: ', -while => $App::moviedb::Movie::fields{title}{re} ) ) {
        say 'Bad value';
    }
    my $movies = $self->{db}->get_movies({ where => { 'movie.title' => $_ } });
    return App::moviedb::Movies->new($movies)->as_string() || 'None found';
}

# find movies by star
sub act_find_by_star {
    my $self = shift;
    while ( ! prompt( 'Star: ', -while => $App::moviedb::Movie::fields{stars}{re} ) ) {
        say 'Bad value';
    }
    my $movies = $self->{db}->get_movies({ where => { 'star.name' => $_ } });
    return App::moviedb::Movies->new($movies)->as_string() || 'None found';
}

# import movies from text file and add to DB
sub act_import {
    my ( $self, $params ) = @_;
    die "Need input file\n" unless $params->{file_in};
    # import movies from text file
    my $movies = $self->_import_txt_file($params->{file_in});
    # add movies to DB
    $self->{db}->add_movies($movies);
    return 'Import '. scalar(@$movies) .' movies done';
}

# list movies by title
sub act_list {
    my $self = shift;
    my $movies = $self->{db}->get_movie_list({ order_by => 'title' });
    return App::moviedb::Movies->new($movies)->as_string() || 'None found';
}

# list movies by year
sub act_list_by_year {
    my $self = shift;
    my $movies = $self->{db}->get_movie_list({ order_by => 'year' });
    return App::moviedb::Movies->new($movies)->as_string() || 'None found';
}

#
# private methods
#

# import movies from text file
# return: arrayref of hashrefs with movie's info
# TODO : create App::moviedb::Import(::Text)
sub _import_txt_file {
    my ( $self, $filename ) = @_;
    use autodie qw( :file );
    # parse file
    my @movies;
    my %fields = (
        'Title'        => 'title',
        'Release Year' => 'year',
        'Format'       => 'format',
        'Stars'        => 'stars',
    );
    my $field_re = qr/[\w\s\-:,]+/;
    open( my $fh, '<', $filename );
    while (<$fh>) {
        chomp();
        next unless $_;
        my $movie;
        if (/^Title:\s*($field_re)$/) {
            $movie->{title} = $1;
            foreach my $field ( ( 'Release Year', 'Format', 'Stars' ) ) {
                my $_ = <$fh>;
                chomp();
                if (/^$field:\s*($field_re)$/) {
                    $movie->{ $fields{$field} } = $1;
                }
                else {
                    die "Bad file format. Expect '$field:' in '$_'\n";
                }
            }
            $movie->{stars} = [ split( /,\s*/, $movie->{stars} ) ];
            push @movies, $movie;
        }
        else {
            die "Bad file format. Expect 'Title:' in '$_'\n";
        }
    }
    close($fh);
    return \@movies;
}

1;

__END__

=head1 NAME

App::moviedb::Command

=head1 SYNOPSIS

	# using inside App::moviedb
    use App::moviedb::Command;
    my $command = App::moviedb::Command->new();
	my $command_method = 'add';
    $command->$command_method(@ARGV) if $command->can($command_method);

=head1 DESCRIPTION

Collection of command/action methods for App::moviedb.

=head1 SEE ALSO

App::moviedb

=head1 AUTHOR

Sergiy Borodych  C<< <bor@cpan.org> >>

=head1 LICENCE AND COPYRIGHT

Copyright (c) 2011, Sergiy Borodych C<< <bor@cpan.org> >>. All rights reserved.

This module is free software; you can redistribute it and/or
modify it under the same terms as Perl itself. See L<perlartistic>.


lib/App/moviedb/DB.pm
package App::moviedb::DB;

use 5.010;
use lib qw( lib );
use warnings;
use strict;

our $VERSION = 0.01;

use DBI;
use DBD::SQLite;

my $db;

sub new {
    my ( $class, $conf ) = @_;
    unless ( $db ) {
        die "Need db driver setup\n" unless $conf->{driver};
        my $self = bless {}, $class;
        $self->{dbh} = DBI->connect(
            'DBI:' . $conf->{driver}
              . ':dbname='
              . ( $conf->{dir} ? "$conf->{dir}/" : '' )
              . $conf->{name}
              . ( $conf->{host} ? ';host=' . $conf->{host} : '' ),
            $conf->{user}, $conf->{password},
            { RaiseError => 1 }
        ) or die "Can't connect to database\nError: $DBI::errstr\n";
        $db = $self;
    }
    return $db;
}

sub dbh {
    my $self = shift;
    return $self->{dbh};
}

# add movies & stars to DB
# return: hashref of araryrefs { star_name => [movie_id,...] }
sub add_movies {
    my ( $self, $movies ) = @_;
    my %stars;
    $self->{dbh}->begin_work;
    my $sth = $self->{dbh}->prepare('INSERT INTO movie (title, year, format) VALUES (?,?,?)');
    foreach my $movie (@$movies) {
        $sth->execute( @$movie{qw(title year format)} );
        my $movie_id = $self->{dbh}->last_insert_id( undef, undef, 'movie', 'movie_id' );
        push @{ $stars{$_} }, $movie_id foreach @{ $movie->{stars} };
    }
    $sth->finish();
    $self->{dbh}->commit;

    # add stars & movie_stars to DB
    $self->{dbh}->begin_work;
    $sth = $self->{dbh}->prepare('INSERT INTO star (name) VALUES (?)');
    my $sth_ms = $self->{dbh}->prepare('INSERT INTO movie_star (movie_id, star_id) VALUES (?,?)');
    foreach my $name ( keys %stars ) {
        $sth->execute($name);
        my $star_id = $self->{dbh}->last_insert_id( undef, undef, 'star', 'star_id' );
        foreach my $movie_id (@{$stars{$name}}) {
            $sth_ms->execute($movie_id,$star_id);
        }
    }
    $sth->finish();
    $sth_ms->finish();
    $self->{dbh}->commit;
    return 1;
}

# delete movie from DB
sub del_movie {
    my ( $self, $id ) = @_;
    $self->{dbh}->do('DELETE FROM movie_star WHERE movie_id = ?', undef, $id);
    my $r = $self->{dbh}->do('DELETE FROM movie WHERE movie_id = ?', undef, $id);
    return $r;
}

# return: 
sub get_movies {
    my ( $self, $params ) = @_;
    my $query = q/
        SELECT DISTINCT movie.movie_id, movie.title
        FROM movie
        JOIN movie_star USING (movie_id)
        JOIN star USING (star_id)
        WHERE 1
    /;
    my @binds;
    foreach ( keys %{ $params->{where} } ) {
        $query .= " AND $_ LIKE ?";
        push @binds, "%$params->{where}{$_}%";
    }
    return $self->{dbh}->selectall_arrayref(
        $query, { Slice => {} }, @binds
    );
}

sub get_movie_list {
    my ( $self, $params ) = @_;
    my $query = 'SELECT movie_id, title FROM movie';
    $query .= " ORDER BY $params->{order_by}" if $params->{order_by};
    return $self->{dbh}->selectall_arrayref( $query, { Slice => {} } );
}

sub get_star {
    my ( $self, $name ) = @_;
    my $star = $self->{dbh}->selectrow_hashref(
        'SELECT star_id, name FROM star WHERE name = ?',
        undef, $name
    );
    return $star;
}

# create db file and connect to it
# return: dbh
sub init_db {
    my $self = shift;
	# TODO
	die "Not implemented yet!\n";
    #my $schema_file = $self->{_conf_dir} . '/moviedb.schema.sql';
    #return $self;
}

1;

__END__

=head1 NAME

App::moviedb::DB

=head1 SYNOPSIS

    use App::moviedb::DB;
    my $dbh = App::moviedb::DB->new();

=head1 DESCRIPTION

This class create dbh singleton.

=head1 AUTHOR

Sergiy Borodych  C<< <bor@cpan.org> >>

=head1 LICENCE AND COPYRIGHT

Copyright (c) 2011, Sergiy Borodych C<< <bor@cpan.org> >>. All rights reserved.

This module is free software; you can redistribute it and/or
modify it under the same terms as Perl itself. See L<perlartistic>.
